Using Val as kernel argument triggers an assertion

MWE
```
using KernelAbstractions

@kernel function kernel_val!(a, ::Val{m}) where {m}
  @inbounds a[1] = m
end

let
  a = zeros(1)
  kernel! = kernel_val!(CPU(), 1, 1)
  event = kernel!(a, Val(1))
  wait(event)
end
```

results in
```
ERROR: LoadError: LoadError: AssertionError: isexpr(decl, :call)
Stacktrace:
 [1] __kernel(::Expr) at /home/mwarusz/repos/KernelAbstractions.jl/src/macros.jl:10
 [2] @kernel(::LineNumberNode, ::Module, ::Any) at /home/mwarusz/repos/KernelAbstractions.jl/src/KernelAbstractions.jl:41
in expression starting at /home/mwarusz/projects/kernels/val_issue.jl:3
in expression starting at /home/mwarusz/projects/kernels/val_issue.jl:3
```

--- a/src/backends/cpu.jl
+++ b/src/backends/cpu.jl
@@ -59,12 +59,12 @@ end
 Cassette.@context CPUCtx
 
 function mkcontext(kernel::Kernel{CPU}, I, _ndrange, _workgroupsize)
-    metadata = CompilerMetadata{workgroupsize(kernel), ndrange(kernel), false}(I, _ndrange, workgroupsize)
+    metadata = CompilerMetadata{workgroupsize(kernel), ndrange(kernel), false}(I, _ndrange, _workgroupsize)
     Cassette.disablehooks(CPUCtx(pass = CompilerPass, metadata=metadata))
 end
 
 function mkcontextdynamic(kernel::Kernel{CPU}, I, _ndrange, _workgroupsize)
-    metadata = CompilerMetadata{workgroupsize(kernel), ndrange(kernel), true}(I, _ndrange, workgroupsize)
+    metadata = CompilerMetadata{workgroupsize(kernel), ndrange(kernel), true}(I, _ndrange, _workgroupsize)
     Cassette.disablehooks(CPUCtx(pass = CompilerPass, metadata=metadata))
 end
 

--- a/src/macros.jl
+++ b/src/macros.jl
@@ -7,7 +7,15 @@ function __kernel(expr)
     body = expr.args[2]
 
     # parse decl
-    @assert isexpr(decl, :call)
+    # `@kernel fname(::T) where {T}`
+    if isexpr(decl, :where) 
+        iswhere = true
+        whereargs = decl.args[2:end]
+        decl = decl.args[1]
+    else
+        iswhere = false
+    end
+    @assert isexpr(decl, :call) 
     name = decl.args[1]
 
     # List of tuple (Symbol, Bool) where the bool
@@ -36,6 +44,11 @@ function __kernel(expr)
 
     gpu_decl = Expr(:call, gpu_name, arglist...)
     cpu_decl = Expr(:call, cpu_name, arglist...)
+
+    if iswhere
+        gpu_decl = Expr(:where, gpu_decl, whereargs...)
+        cpu_decl = Expr(:where, cpu_decl, whereargs...)
+    end
 
     # Without the deepcopy we might accidentially modify expr shared between CPU and GPU
     gpu_body = transform_gpu(deepcopy(body), args)

--- a/test/test.jl
+++ b/test/test.jl
@@ -168,3 +168,17 @@ end
         end
     end
 end
+
+@kernel function kernel_val!(a, ::Val{m}) where {m}
+    I = @index(Global)
+    @inbounds a[I] = m
+end
+
+A = zeros(Int64, 1024)
+wait(kernel_val!(CPU())(A,Val(3), ndrange=size(A)))
+@test all((a)->a==3, A)
+if has_cuda_gpu()
+    A = CuArrays.zeros(Int64, 1024)
+    wait(kernel_val!(CUDA())(A,Val(3), ndrange=size(A)))
+    @test all((a)->a==3, A)
+end

